[WIP] feat: proyectos confirmados

Pablom25 opened the pull request:
Añadir posibilidad de que haya proyectos confirmados y proyectos sin confirmar

Pablom25 committed:
first commit

Pablom25 committed:
ref .gitignore

Pablom25 committed:
feat: añadir datos, proyectos confirmados y a optimizar

Pablom25 committed:
feat: devuelve set proyectos confirmados y sin confirmar, y diccionario de longitud ocupada/localización por proyectos confirmados

Pablom25 committed:
feat: actualizar funciones a proyectos confirmados

Pablom25 committed:
feat: lo mismo pero en .ipynb

MonicaCuadrado commented on leerDatos.py:
```diff
@@ -18,18 +18,20 @@ def leer_datos() -> tuple[pd.DataFrame, pd.DataFrame, pd.DataFrame, pd.Timestamp
+        'eslora': [120, 100, 120],
+        'manga': [18, 15, 18],
+        'proyecto_id': ['PRO1', 'PRO2', 'PRO3'],
+        'nombre_area': ['sin_ubicacion_asignada', 'sin_ubicacion_asignada', 'SUR'],
```
para seguir nuestra misma nomenclatura llamar: "SIN UBICACION ASIGNADA" y "MUELLE SUR" "MUELLE NORTE"

Pablom25 replied:
Dentro del df de muelles también los llamo "MUELLE SUR" y "MUELLE NORTE" o lo dejo como "SUR" y "NORTE"?

MonicaCuadrado replied:
tambien con MUELLE

MonicaCuadrado commented on procesarDatos.py:
```diff
@@ -33,4 +39,11 @@ def preprocesar_datos(proyectos: pd.DataFrame, periodos: pd.DataFrame, muelles:
+    set_confirmados = set(proyectos[~proyectos['proyecto_a_optimizar']].index)
+
+    # Crear diccionario de longitud total de barcos confirmados por ubicación
+    longitudes_confirmados = proyectos[~proyectos['proyecto_a_optimizar']].groupby('nombre_area')['eslora'].sum().to_dict()
```
los confirmados ocupan cierta área **en cierto día**

MonicaCuadrado commented on optimize.py:
```diff
@@ -96,16 +100,17 @@ def definir_restricciones(x: dict, y: dict, dias: list, dias_vars: dict, locs_va
     restricciones.update(
         {
-            f"Longitud_Muelle_{d}_{loc}": (pulp.lpSum(x.get((p, d, loc),0) * proyectos.loc[periodos.loc[p, 'proyecto_id'], 'eslora'] for p in periodos.index) <= muelles.loc[loc, 'longitud'], 
+            f"Longitud_Muelle_{d}_{loc}": (pulp.lpSum(x.get((p, d, loc),0) * proyectos.loc[periodos.loc[p, 'proyecto_id'], 'eslora'] for p in periodos.index) + longitudes_confirmados.get(loc,0) <= muelles.loc[loc, 'longitud'], 
```
cuidado: aquí estás considerando que los confirmados ocupan metros todos los días, no solo los días que realmente están en esa ubi

Pablom25 committed:
ref: cambio nomenclatura ubicacion

Pablom25 committed:
ref: cambiar nombre muelles

Pablom25 committed:
fix: longitudes_confirmados arreglado

- Se calcula por dia y ubicación
- Ahora calculado en la misma función de restricciones (no se usa en ningún otro lado)

Pablom25 committed:
feat: lo mismo pero en un .ipynb

MonicaCuadrado commented on optimize.py:
```diff
@@ -95,6 +95,14 @@ def definir_restricciones(x: dict, y: dict, dias: list, dias_vars: dict, locs_va
         Diccionario de restricciones del problema de optimización.
     """
 
+    # Crear diccionario de longitud total de barcos confirmados por ubicación y por dia
+    longitudes_confirmados = {}
+    for d in dias:
+        for m in muelles.index:
+            for p in set_confirmados:
+                if d >= periodos.loc[periodos['proyecto_id']==p]['fecha_inicio'].values[0] and d <= periodos.loc[periodos['proyecto_id']==p]['fecha_fin'].values[0] and m == proyectos.loc[p, 'nombre_area']:
+                    longitudes_confirmados[(d,m)] = longitudes_confirmados.get(d, 0) + proyectos.loc[p, 'eslora']
+
```
en el momento que se llegan a estas complejidades de triple bucle for es recomendable definir en lugar de diccionarios nuevas columnas en el específico df. Por ejemplo en este caso podrías tener en el df periodos:
* una columna llamada `dias`: que contenga la lista de dias donde está definido cada periodo
* una columna llamada `ubicaciones` que contenga la lista de ubicaciones donde puede ser asignado el proyecto en dicho periodo

Una vez que las tienen jugar con el explode() y el groupby() hace que el código sea más limpio y claro

Pablom25 committed:
ref: crear columna dias y ubicaciones

- Ahora en procesar datos en vez de en definir variables
- Ahora como parte del df en vez de como diccionarios

Pablom25 committed:
ref: longitudes_confirmados con merge, explode y groupby

MonicaCuadrado commented on optimize.py:
```diff
@@ -96,22 +81,18 @@ def definir_restricciones(x: dict, y: dict, dias: list, dias_vars: dict, locs_va
-            for p in set_confirmados:
-                if d >= periodos.loc[periodos['proyecto_id']==p]['fecha_inicio'].values[0] and d <= periodos.loc[periodos['proyecto_id']==p]['fecha_fin'].values[0] and m == proyectos.loc[p, 'nombre_area']:
-                    longitudes_confirmados[(d,m)] = longitudes_confirmados.get(d, 0) + proyectos.loc[p, 'eslora']
+    mergeddf = periodos.merge(proyectos[['eslora', 'nombre_area']], left_on='proyecto_id', right_index=True, how='left')
```
por qué proyectos tiene una columna llamada `nombre_area`? ese atributo lo tiene asociado cada periodo de repración asociado al proyecto. No es a nivel de proyecto.

MonicaCuadrado replied:
Si lo que quieres es añadir la columna "eslora" a periodos, hazlo con un map sobre "proyecto_id"

Pablom25 committed:
fix: mover nombre_area a periodos

Pablom25 committed:
ref: cambiar nomenclatura sinconfirmar/confirmados a a_optimizar/no_optimizar

Pablom25 committed:
perf: map en vez de merge

Pablom25 committed:
ref: lo mismo pero en el .ipynb

Pablom25 committed:
fix: arreglar imprimir asignación
